opacus/privacy_engine.py

```diff
@@ -31,15 +31,14 @@ class PrivacyEngine:
         >>> import torch
         >>> model = torch.nn.Linear(16, 32)  # An example model
         >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.05)
-        >>> privacy_engine = PrivacyEngine(model, batch_size, sample_size, alphas=range(2,32), noise_multiplier=1.3, max_grad_norm=1.0)
+        >>> privacy_engine = PrivacyEngine(model, sample_rate, alphas=range(2,32), noise_multiplier=1.3, max_grad_norm=1.0)
         >>> privacy_engine.attach(optimizer)  # That's it! Now it's business as usual.
     """
 
     def __init__(
         self,
         module: nn.Module,
-        batch_size: int,
-        sample_size: int,
+        sample_rate: float,
```

Comment:
Haha this is a point of contention indeed, and something that we kept going back to. If we follow the Abadi et al convention, then yes we should use `sample_rate` in lieu of `batch_size` and `len(dataset)`. On the other hand, PyTorch really treats the latter convention as a first-class citizen, and it shows everywhere (for example, you design models thinking about your batch dimension, not a sample rate). So the decision was to make the API more PyTorch-friendly, given that the sample rate can easily be computed from those two anyway.

Curious to hear your perspective on this

Reply:
It is  exactly what I would like to fix with this PR. The main bug in Opacus library is the fact that the batch construction from Abadi et al [2] is used with the privacy budget evaluation from Mironov et al [1]. This privacy budget evaluation [1] is based on the Sampled Gaussian Mechanism. As a consequence, this evaluation assumes that to build one batch, each sample in the dataset is selected with probability q (called sample rate). With this batch construction, the batch size does not exist because the batches will have variable size. Using
another batch construction will break this assumption and the privacy budget evaluation based on Mironov et al [1] will be incorrect (as pointed out by Nasr et al in [3]).

To fix this bug, we have two options:
- use the batch construction and the privacy budget evaluation from Abadi et al [2]. With this privacy budget evaluation based on moment accountants, we will obtain a higher privacy budget than the current privacy budget implemented in Opacus. This first option implies to re-implement the overall privacy budget evaluation.
- use the batch construction and the privacy budget evaluation from Mironov et al [1]. For this option, we have just to change the parameters of the PrivacyEngine (the batch size does not exist with this batch construction) and we have to update all examples and tutorials in order to use batches sampled according to the Sampled Gaussian Mechanism.

[1] Rényi Differential Privacy of the Sampled Gaussian Mechanism
[2] Deep Learning with Differential Privacy
[3] Improving deep learning with differential privacy using gradient encoding and denoising

Reply:
... Of course, the batch size is indeed variable. Sorry, my bad!

Reply:
Indeed, this question now doesn't make sense in light of the variable batch size. 

I agree to add this capability, but I want to make sure we don't introduce a breaking change all of a sudden as this is an API change.

How about we do the following? 

1. We leave batch_size and sample_size as they are. In an ideal world we would make them optional, but this would mean moving them after `noise_multiplier` and introduce yet another breaking change.
2. We add an optional parameter `dataloader = None`.
3. If the dataloader is not `None`, we print a warning that it will override the `batch_size` and the `sample_size` settings you provided.
4. Check if the dataloader has a sample_rate. If not, raise another warning that`opacus.utils. UniformWithReplacementSampler` should be used, and refer to the examples to see how to use it.

We are working on a major API refactor for Opacus 1.0, so in the future we will make this better, but this should suffice for now.

Reply:
I have chosen to put batch_size, sample_size and sample_rate arguments optional